### DISCUSSION TITLE
Resource deletion - consider support for cascading forward delete and blockOwnerDeletion?

### DISCUSSION THREAD

**Discussion by bobh66** [crossplane/crossplane#3116] [General]
I'd like to start a discussion about the way composite/managed resource deletion is handled and some of the problems that seem to be pretty common,  I don't pretend to be an expert in this area but I believe a solution to this is required before Crossplane can be truly useful at scale.

There are a fair number of issues across different projects that are all related to the way Crossplane handles delete processing:

#1612 
https://github.com/crossplane/crossplane/issues/2439
https://github.com/crossplane-contrib/provider-helm/issues/115
https://github.com/crossplane-contrib/provider-kubernetes/issues/38
https://github.com/crossplane-contrib/provider-terraform/issues/47

The common thread through most of these issues is that the current delete implementation ignores the implicit dependencies between the infrastructure that is being created (usually Clusters of some type, but also other cloud resources) and the "next layer(s)" of things/resources/services that are being provisioned into/onto that infrastructure.

When the top level Claim is deleted, kubernetes dutifully traverses the ownership tree and triggers deletion of all of the associated composite and managed resources.  This is the functional equivalent to "pulling the rug out" from under all of the resources that have been created.  The Composite resources go away pretty much immediately, most of the Managed resources eventually get deleted and go away, but depending on the content of the Composition there can be several or many Managed resources (helm Releases, kubernetes Objects, terraform Workspaces) that are simply stuck and can never be reconciled or deleted.  All three of these resource types depend on ProviderConfigs which depend on Secrets which point at Clusters or other things which may or may not have been deleted already.  There may also be additional cloud resources that were created outside of Crossplane as a result of a Crossplane-managed object - for example an AWS ELB that gets created because an nginx/ingress was created.

Since the Clusters and Secrets created by the Composition can be deleted before the Objects/Releases/Workspaces that rely on them, it would seem that there are only a few possible solutions:

- Update the Providers to handle this scenario by allowing the Observe and Delete processing to complete "successfully" if the provider no longer has valid credentials for the target Cluster AND the resource has been marked for deletion.  This raises the possibility that the resource has been marked for deletion and access to the remote cluster is unavailable for some reason other than the cluster has also been deleted, leaving the remote version of the resource orphaned on the cluster.  However today we have orphaned Object/Release/Workspace objects that consume resources and reconciliation time in the providers when the remote clusters that they point at are long deleted and will never be accessible.  The result is that these providers will continually slow down and eventually become unusable as they spend more and more time reconciling orphaned resources against remote clusters that no longer exist.
This does not resolve the problem completely, as there may be orphaned cloud resources that would only get deleted if the Object/Release/Workspace was properly deleted before the Cluster went away.

- Use a combination of cascading forward deletion and the blockOwnerDeletion flag to cause a "bottom up" deletion path so that child resources are deleted before their owners.  Crossplane already adds ownerReferences to all of the composite and managed resources, though it does not set the blockOwnerDeletion flag, and Crossplane also removes any ownerReferences that may have been added via patches.   If Crossplane were to set the blockOwnerDeletion flag on these ownerReferences, when a Claim is deleted with --cascade=forward a "forwardDeletion" finalizer will be added to every object when it is marked deleted, and the owner objects will not be deleted until their children have been deleted.

Crossplane and crossplane-runtime could be modified to detect the forwardDeletion finalizer and delay the deletion of the associated composite/managed resources until the forwardDeletion finalizer has been removed.  That would allow for a "bottom-up" deletion sequence which gives the "application" layer resources a chance to finish deletion cleanly before the underlying infra is removed.

I know there is a reluctance to support terraform-like dependency specification in the Composition, and I understand the reasoning, but there is an implicit dependency map created by the resources defined in the composition, or set of compositions, and Crossplane is already explicitly specifying ownerReferences, so it seems logical to leverage that to strengthen the implicit/explicit dependencies and use kubernetes forward deletion to manage the deletion path.

I'm interested in opinions/ideas/suggestions about how best to resolve this.  As it is now we have to do a lot of manual cleanup. or write scripts to do cleanup, and it's not sustainable as our Compositions and combinations of Compositions get more complex.

**Comment by muvaf** [comment]
Thanks for a great write-up @bobh66 !

> Update the Providers to handle this scenario by allowing the Observe and Delete processing to complete "successfully" if the provider no longer has valid credentials for the target Cluster AND the resource has been marked for deletion. This raises the possibility that the resource has been marked for deletion and access to the remote cluster is unavailable for some reason other than the cluster has also been deleted, leaving the remote version of the resource orphaned on the cluster.

I share your concern here that making the general connection failure a valid path is risky and very specific to this use case. We wouldn't really want to do this change in the managed reconciler and having that behavior only in select CRDs is surprising for users.

> Use a combination of cascading forward deletion and the blockOwnerDeletion flag to cause a "bottom up" deletion path so that child resources are deleted before their owners.

I'm not sure I follow this option. How would this somehow block deletion of the `Cluster` (managed resource, GKE Cluster, EKS cluster etc) while the resources provisioned in them by provider-helm or provider-kubernetes are being deleted?


I'd like to write a short summary of my understanding of the problem before making suggestions. We have a `ProviderConfigUsage` type that prevents `ProviderConfig` from being deleted. Let's say you have a `Cluster` and a `ProviderConfig` that points to its connection secret. In this case, `provider-helm` or `provider-kubernetes` resources depending on that `ProviderConfig` are able to keep it from being deleted just by referencing it in `spec.providerConfigRef.name` because `ProviderConfigUsage` is automatically created once the provisioning using that `ProviderConfig` is started. The problem seems to be that this does not block the deletion of the actual `Cluster`, so we end up having `ProviderConfig`, `ProviderConfigUsage` and say `Release` object from `provider-helm` but there is no `Cluster` since it got the deletion signal and started the deletion process. Is my understanding correct?

Assuming this is the case, the main culprit is that the `ProviderConfig` points to the `Secret` of the `Cluster` and that's pretty much all is known about the relationship and it's more of a passive one since neither `Secret` nor `Cluster` is aware that something points to them in the API. I wonder if we can have this consumption represented at the API level similar to `ProviderConfigUsage`. Three options coming to my mind:
* We could possibly do this by having a `ConnectionSecretUsage` type that will block the deletion of the `Secret` of `Cluster` and by changing the code [here](https://github.com/crossplane/crossplane-runtime/blob/86fb15d/pkg/reconciler/managed/reconciler.go#L823) in the managed reconciler, we can make it such that if the connection secret cannot be deleted, then the external resource, i.e. `Cluster`, deletion is blocked.
  * Instead of a new type, we could make `ProviderConfig` controller add a finalizer to the `Secret` it references as long as there are `ProviderConfigUsage`s in place. It'd remove the finalizer once there is no `ProviderConfigUsage` in place.
* We could be more explicit that we have users specify which resource the `ProviderConfig` points to under its `spec` in an array; essentially an arbitrary dependency mechanism but only on `ProviderConfig`.

I feel like the first one requires less intervention by defaulting to block and it's pretty specific. From the perspective of the controller of `Cluster` resource, it does feel a bit weird that the `Secret` blocks the deletion but with all the use cases we've seen, I'm getting used to the idea more and more. What do you think?

**Reply by bobh66** [reply]
@muvaf thanks for your response and sorry for the delayed reply.

Your understanding of the problem is correct, I'm just not sure that addressing the problem from the ProviderConfig perspective will be practical for all of the different scenarios.

First let me add the one critical piece that my original description was missing - the ability for the user (or anything really - a policy engine?) to set ownerReferences when creating a resource in a Composition.  If the user had the ability to set the ownerReferences for a resource to a logical "parent" resource then by using a combination of blockOwnerDeletion and cascaded forward delete the deletion graph can be influenced enough to prevent the problems we see with orphaned resources.   Today the Crossplane composite code removes all user-provided ownerReferences as a precaution against someone else taking over as the controller, which is understandable.  If this restriction can be eased slightly to (forcibly) prevent any other ownerReference from having controller: true, but still allow the existence of other "owners", along with the ability to set blockOwnerDeletion: true, that would go a long way towards providing support for cascaded forward deletion.

It would be interesting to see if it's possible to write Kyverno policies that detect dependency relationships between different composite/managed resources and add ownerReferences that represent those dependencies.  That would automate the dependency tree required for cascaded forward delettion.

At the infrastructure level these dependencies are handled by the providers - you can't delete a VPC while it has Subnets and you can't delete Subnets while they have ENIs and you can't delete the ENIs while the EC2 instances are running, and that all gets handled by reconciliation.  Eventually the EC2 instances terminate and the ENIs get deleted and the Subnets get deleted and the VPC goes away.

The application layer adds another level of complexity that the infrastructure is unaware of and it can't be handled by the infrastructure providers.  Helm Releases should be able to be deleted cleanly, which means they need to be able to run pre-delete and post-delete jobs, which means the associated NodeGroup(s) can't go away before the Release so they can execute the jobs, and the Cluster can't go away before the Nodegroup, and the Connection Secret can't go away before the Cluster, and the ProviderConfig can't go away before the Secret.  And that's just one scenario for one cloud provider.  The same thing applies for Kubernetes Objects and for Terraform Workspaces that rely on the cluster being there for the delete handling.  When the Cluster/Nodegroup/Secret/etc all go away before these resources have been deleted they get hung and need to be cleaned up so they don't waste CPU resources in the provider.

I don't think it's practical to track the dependencies backwards from the ProviderConfig to the Secret to the Cluster to the Nodegroup, and there may be other scenarios with other dependencies that these changes would not cover.

One other possibility is to just set deletionPolicy: Orphan and let the managed resource go away without worrying about it's corresponding external resource.  That works fine when deploying the entire infrastructure at once - if the cluster is going away anyway, who cares what happens to the Helm Releases and the Kubernetes objects?  But if something needs to be deleted and redeployed into an existing Cluster, or modified in an existing deployment, it can't be marked as Orphan - it needs to be able to be deleted properly and re-instantiated.

The more I think about it using Kyverno to define the dependencies automatically - and keeping this functionality out of Crossplane - would be an ideal solution.

If Crossplane can:
- allow additional (non-controller) ownerReferences 
- support "external" finalizers by delaying Delete processing until they go away

it opens up all kinds of possibilities for external applications to manipulate Crossplane resources however they need to.

I'd be happy to do a PoC of this if these ideas are acceptable.

Thanks!

**Reply by nabuskey** [reply]
I agree that adding dependency specification in Crossplane is not a good idea and should be handled outside of it. However, I am not quite understanding why you want to use Kyverno for this. It's likely possible to do this with Kyverno, but it feels like it's operating outside of its concerns. I think infrastructure and application dependency specification should be handled by your deployment tooling of choice. ArgoCD and Flux both offer functionalities for this kind of scenarios.

**Comment by bobh66** [comment]
I don't see Kyverno as having a particular area of concern - I think it is a very flexible tool that can be used for whatever a user decides they need it for, enforcing all sorts of different policies as needed.  In this case mapping standard dependencies between resources could be done fairly easily, which eliminates complexity in the composition by centralizing the dependency rules in a set of policies.  I'm sure there are other solutions, and Flux/ArgoCD may be better in some scenarios and not in others.  The important part is to make sure that the user has the ability to make those choices and that Crossplane doesn't enforce limitations that aren't absolutely required.

**Comment by bobh66** [comment]
#3211 has been opened for the ownerReference restriction issue

**Comment by bobh66** [comment]
After some more research and testing, it's clear that foreground cascading deletion on it's own is not a complete solution, as it results in all of the "leaf" resources being marked deleted, with no foregroundDeletion finalizer, regardless of the dependencies of the higher level composites.  In addition to the changes described above it's necessary to manage dependencies on composite resources using finalizers on the entire tree of dependencies to prevent deletion of any of the component resources of the composite before the dependency on the composite is removed.

**Comment by bobh66** [comment]
Adding a pointer to a Slack discussion https://crossplane.slack.com/archives/C03U14KRK6H/p1663951842605469?thread_ts=1663809273.781329&cid=C03U14KRK6H

And some additional questions:

1 - Does Crossplane want to be responsible for / facilitate dependency management in the DSL/API? (I think the answer to date has been "no")
2 - Does Crossplane want to allow users to manage their own dependencies via existing Kubernetes mechanisms such as ownerReferences and finalizers?

If the answer to (2) is yes, then assuming both Foreground Cascading Deletion is supported (#3225), and finalizers are cascaded to the edge nodes (#3323), a Composition owner can define their own dependencies using the existing Crossplane DSL and a combination of provider-kubernetes dependsOn and ownerReferences.

For example, given a composite resource A and another Composite resource B, deployed in the same composition, where B needs to be deleted before A.

We can use provider-kubernetes to deploy Composite resource B inside an Object and set a dependsOn reference to A, along with an ownerReference in B that points at A with blockOwnerDeletion set to true.  That will cause provider-kubernetes to set a finalizer on A, which gets reconciled out to the edges and prevents all of A's resources from being deleted while the finalizer exists.

The ownerReference from Object/B to A causes the kubernetes Foreground Cascading Deletion code to require that all of B's resources are deleted before A can be deleted.  When B is finished deleting, the provider-kubernetes finalizer gets removed from A and all of A will be deleted.

This allows the composition authors to specify the dependencies that they know about and handle delete order issues without major changes to the Crossplane.

**Comment by bobh66** [comment]
Tagging @negz as he is looking at this as well

**Comment by bobh66** [comment]
Proposal to support resource usage for dependency/delete order - https://github.com/crossplane/crossplane/issues/3393

**Comment by bobh66** [comment]
I have a PoC-grade implementation of delete dependencies using Kyverno policies - working on a writeup

**Reply by jcooklin** [reply]
@bobh66 will you point me to the writeup or the kyverno policies you used?  🙏

**Comment by bobh66** [comment]
@jcooklin https://github.com/crossplane/crossplane/discussions/4072
